### PR TITLE
[Feature #159053986] Add endpoints and functionality for creating articles

### PR DESCRIPTION
**Pull request by hoslack**
**What does this PR do?**
Implements the creating of articles by the users

**Description of Task to be completed?**
- Add a new app called articles
- Create models, serializers and views for articles and migrate models to the database
- Create endpoints that allow the creation of articles

**How should this be manually tested?**
- Clone the repository from [here](https://github.com/andela/ah-scorpion) and check out the branch `ft-159053986-Add-endpoint-for-Creating-Posts`
- Activate your virtual environment and use the `.env-sample` to set the environment variables.
- Use postman to test these endpoints.
  `POST /articles/` for creating an article

{
  
    "title": "How to train your dragon",
    "description": "Ever wonder how?",
    "body": "You have to believe",
    "tagList": ["reactjs", "angularjs", "dragons"],
    "images" : ["imagelink1", imagelink2]
}
- **You need to login first to use the endpoint. Testing without loging in will give an error**

What are the relevant pivotal tracker stories?
[#159053986](https://www.pivotaltracker.com/n/projects/2184741/stories/159053986)

### CODE REVIEW

**Review by RuthNjeri — COMMENTED** [review]
Add tests for creating articles with its edge cases such as a user who is not an author writing an article.

**Comment by dmigwi** [comment]
Your last commit looks like It was initially a merge conflict. Please avoid making such commits, you should rebase your branch with develop.

**Review by davenjeru — CHANGES_REQUESTED** [review]
I have tested this and:
1. I cannot seem to create an article
2. The endpoint for creating an article should only allow authenticated users

File: authors/urls.py [hunk 1]
```diff
@@ -18,9 +18,9 @@
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
-
-    url(r'^api/', include('authors.apps.authentication.urls',
-                          namespace='authentication')),
+    url(r'^api/', include('authors.apps.authentication.urls', namespace='authentication')),
+    url(r'^articles/', include('authors.apps.articles.urls', namespace='articles')),
```
Comment (munyili-samuel) [inline authors/urls.py]: Any reason why you omitted the `api/` part from the articles url? I think that should be included in all urls

**Comment by munyili-samuel** [comment]
When I follow the instructions to test the feature I get the following JSON response:
```
{
    "detail": "Expired Token."
}
```
Does that mean that I should log in first? If so I'd recommend that you add that one needs to login first

File: authors/urls.py (hunk 1 above)
Comment (hoslack) [reply authors/urls.py]: Done

**Comment by dmigwi** [comment]
@hoslack You've so many migration versions of the same field whereas this is the first implementation of the articles features.
screen shot 2018-08-02 at 19 22 31

Drop the migrations in the articles folder and generate new migrations. That should just create a single migration file instead of the seven we have now in your PR.

File: authors/apps/articles/views.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+from django.http.response import Http404
+from django.utils.text import slugify
+from rest_framework.response import Response
+from rest_framework import status
+from .serializers import ArticleSerializer
+from .models import Article
+
+
+class ArticleList(APIView):
+    """
```
Comment (dmigwi) [inline authors/apps/articles/views.py]: If you used generic views on this page you could have only a maximum of ten lines of code excluding the imports. You need to implement something like this here http://www.django-rest-framework.org/tutorial/3-class-based-views/#using-generic-class-based-views

**Review by dmigwi — COMMENTED** [review]
Your work looks great though. Fix the migration thing now and we shall modify the `views.py` file tomorrow in the morning.

**Review by davenjeru — APPROVED** [review]

**Review by RuthNjeri — DISMISSED** [review]
You have too many migration files in the ./articles/migrations folder as mentioned by @dmigwi.
You can delete them and leave only one.

**Review by lennykioko — APPROVED** [review]
Good job. Will notify you if I notice anything later on.

**Review by RuthNjeri — APPROVED** [review]

File: authors/apps/authentication/models.py [hunk 3]
```diff
@@ -118,3 +118,6 @@ def get_short_name(self):
         the user's real name, we return their username instead.
         """
         return self.username
+
```
Comment (dmigwi) [inline authors/apps/authentication/models.py:121]: remove this new lines.

**Review by dmigwi — CHANGES_REQUESTED** [review]
Fix that first.

**Review by RuthNjeri — APPROVED** [review]

**Comment by dmigwi** [comment]
Great work for exceeding expectations. Your task needed you to only implement the create articles functionality only but you went ahead and implemented the whole CRUD functionality for the articles.

Unfortunately only `"GET /api/articles/1/ HTTP/1.1" 500 16293` is failing. Please fix it and I will go ahead to merge the PR.

Also remember to fix the merge conflicts in this PR

File: authors/apps/articles/tests/test_views.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        """
+        Test that a user can't create an article without authorisation.
+        """
+        response = self.client.post('/api/articles/', self.data, format='json')
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    def test_create_article_successfully(self):
+        """
+        Test that user can't create article without data
+
```
Comment (lennykioko) [inline authors/apps/articles/tests/test_views.py]: Is this extra whitespace in the docstrings necessary?

File: authors/apps/articles/serializers.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+from .models import Article
+
+
+class ArticleSerializer(serializers.ModelSerializer):
+    slug = serializers.CharField(
+        max_length=100,
+        validators=[UniqueValidator(
+            queryset=Article.objects.all(),
+            message="Title already exists, please enter a different Title")],
```
Comment (lennykioko) [inline authors/apps/articles/serializers.py:12]: I think this error message is confusing given that you say title already exists when referring to the slug.

Comment (hoslack) [reply authors/apps/articles/serializers.py:12]: The title is what the user enters. Not the slug! We create the slug from the title
@lennykioko

File: authors/apps/articles/tests/test_views.py (hunk 4 above)
Comment (hoslack) [reply authors/apps/articles/tests/test_views.py]: White spaces removed @lennykioko

File: authors/apps/articles/models.py [hunk 6]
```diff
@@ -0,0 +1,27 @@
+from django.contrib.postgres.fields import ArrayField
+from django.db import models
+
+from ..authentication.models import User
+
+
+class Article(models.Model):
+    """This is a model for storing articles in the database"""
+    slug = models.CharField(max_length=200, null=True, blank=True, unique=True)
```
Comment (lennykioko) [inline authors/apps/articles/models.py:9]: Could we have this as a Slugfield?

Comment (hoslack) [reply authors/apps/articles/models.py:9]: We agreed to do it in the next story

**Review by dmigwi — APPROVED** [review]
